File: dcds-flags/scripts/export-figma.py
#!/usr/bin/env python3
"""
Export flag SVGs from Figma.
Usage: FIGMA_TOKEN=your_token python3 scripts/export-figma.py
"""

import json
import os
import sys
import time
import urllib.request
from pathlib import Path

FIGMA_TOKEN = os.environ.get("FIGMA_TOKEN")
if not FIGMA_TOKEN:
    print("Error: FIGMA_TOKEN is required", file=sys.stderr)
    sys.exit(1)

FILE_KEY = "P52nmDshYaKr963q1zBwQj"
BATCH_SIZE = 100
SCRIPT_DIR = Path(__file__).parent
MANIFEST_PATH = SCRIPT_DIR / "flags-manifest.json"
OUT_DIR = SCRIPT_DIR.parent / "flags"

OUT_DIR.mkdir(exist_ok=True)

manifest = json.loads(MANIFEST_PATH.read_text())
flags = [f for f in manifest if f.get("nodeId")]

print(f"Exporting {len(flags)} flags from Figma...")

saved = 0
errors = 0

def figma_request(url):
    req = urllib.request.Request(url, headers={"X-Figma-Token": FIGMA_TOKEN})
    with urllib.request.urlopen(req) as r:
        return json.loads(r.read())

def download(url, dest, retries=3):
    for attempt in range(retries):
        try:
            urllib.request.urlretrieve(url, dest)
            return True
        except Exception as e:
            if attempt == retries - 1:
                return False
            time.sleep(attempt + 1)

total_batches = (len(flags) + BATCH_SIZE - 1) // BATCH_SIZE

for batch_num, i in enumerate(range(0, len(flags), BATCH_SIZE), 1):
    batch = flags[i:i + BATCH_SIZE]
    node_ids = ",".join(f["nodeId"].replace(":", "-") for f in batch)
    print(f"  Batch {batch_num}/{total_batches} ({len(batch)} flags)... ", end="", flush=True)

    try:
        url = (
            f"https://api.figma.com/v1/images/{FILE_KEY}"
            f"?ids={node_ids}&format=svg&svg_include_id=false&svg_simplify_stroke=true"
        )
        data = figma_request(url)
        images = data.get("images", {})

        for flag in batch:
            svg_url = images.get(flag["nodeId"])
            if not svg_url:
                print(f"\n  ⚠ No URL for {flag['code']}")
                errors += 1
                continue

            dest = OUT_DIR / f"{flag['code']}.svg"
            if download(svg_url, dest):
                saved += 1
            else:
                print(f"\n  ⚠ Download failed for {flag['code']}")
                errors += 1

        print("done")

    except Exception as e:
        print(f"FAILED: {e}")
        errors += len(batch)

    if i + BATCH_SIZE < len(flags):
        time.sleep(0.3)

print(f"\n✓ Export complete: {saved} SVGs saved, {errors} errors")

File: dcds-flags/scripts/export-figma.mjs
#!/usr/bin/env node
/**
 * Exports flag SVGs from Figma using the REST API.
 *
 * Usage:
 *   FIGMA_TOKEN=your_token node scripts/export-figma.mjs
 *
 * Flags are exported to ../flags/{CODE}.svg
 * Node IDs must be filled in scripts/flags-manifest.json before running.
 */

import { mkdir, writeFile, readFile } from 'fs/promises'
import path from 'path'
import { fileURLToPath } from 'url'

const __dirname = path.dirname(fileURLToPath(import.meta.url))
const ROOT = path.join(__dirname, '..')

const FIGMA_TOKEN = process.env.FIGMA_TOKEN
const FILE_KEY = 'P52nmDshYaKr963q1zBwQj'
const BATCH_SIZE = 100

if (!FIGMA_TOKEN) {
  console.error('Error: FIGMA_TOKEN is required')
  console.error('  Usage: FIGMA_TOKEN=your_token node scripts/export-figma.mjs')
  process.exit(1)
}

const manifest = JSON.parse(
  await readFile(path.join(__dirname, 'flags-manifest.json'), 'utf-8')
)

const missing = manifest.filter(f => !f.nodeId)
if (missing.length > 0) {
  console.warn(`⚠ ${missing.length} entries have no nodeId — fill in flags-manifest.json first:`)
  missing.forEach(f => console.warn(`   ${f.code} (${f.name})`))
  const ready = manifest.filter(f => f.nodeId)
  if (ready.length === 0) process.exit(1)
  console.warn(`Continuing with ${ready.length} ready entries...\n`)
}

const flags = manifest.filter(f => f.nodeId)

async function exportBatch(nodeIds) {
  const ids = nodeIds.map(id => id.replace(':', '-')).join(',')
  const url = `https://api.figma.com/v1/images/${FILE_KEY}?ids=${ids}&format=svg&svg_include_id=false&svg_simplify_stroke=true`

  const res = await fetch(url, {
    headers: { 'X-Figma-Token': FIGMA_TOKEN },
  })

  if (!res.ok) {
    const text = await res.text()
    throw new Error(`Figma API ${res.status}: ${text}`)
  }

  const { images, err } = await res.json()
  if (err) throw new Error(`Figma export error: ${err}`)
  return images
}

async function downloadSvg(url, retries = 3) {
  for (let i = 0; i < retries; i++) {
    try {
      const res = await fetch(url)
      if (!res.ok) throw new Error(`HTTP ${res.status}`)
      return res.text()
    } catch (e) {
      if (i === retries - 1) throw e
      await new Promise(r => setTimeout(r, 1000 * (i + 1)))
    }
  }
}

await mkdir(path.join(ROOT, 'flags'), { recursive: true })

let saved = 0
let errors = 0

for (let i = 0; i < flags.length; i += BATCH_SIZE) {
  const batch = flags.slice(i, i + BATCH_SIZE)
  const nodeIds = batch.map(f => f.nodeId)
  const batchNum = Math.floor(i / BATCH_SIZE) + 1
  const totalBatches = Math.ceil(flags.length / BATCH_SIZE)

  process.stdout.write(`Batch ${batchNum}/${totalBatches} (${batch.length} flags)... `)

  try {
    const images = await exportBatch(nodeIds)

    await Promise.all(
      batch.map(async (flag, j) => {
        const nodeId = nodeIds[j].replace(':', '-')
        const svgUrl = images[nodeId]

        if (!svgUrl) {
          console.warn(`\n  ⚠ No URL for ${flag.code}`)
          errors++
          return
        }

        try {
          const svg = await downloadSvg(svgUrl)
          await writeFile(path.join(ROOT, 'flags', `${flag.code}.svg`), svg, 'utf-8')
          saved++
        } catch (e) {
          console.warn(`\n  ⚠ ${flag.code}: ${e.message}`)
          errors++
        }
      })
    )

    console.log('done')
  } catch (e) {
    console.error(`FAILED: ${e.message}`)
    errors += batch.length
  }

  if (i + BATCH_SIZE < flags.length) {
    await new Promise(r => setTimeout(r, 300))
  }
}

console.log(`\n✓ Export complete: ${saved} SVGs saved, ${errors} errors`)
